projects/bootloader/scripts/can_datagram.py

```diff
@@ -0,0 +1,343 @@
+                self.datagram_messages.append(byte)
+            self.receiving_datagram = True
+
+        elif msg.arbitration_id == 0x0000 and self.receiving_datagram:
+            # Here we want to handle what happens when we get the next message in a datagram
+            message_bytearray = msg.data
+
+            # We handle if there are still node ids in this message
+            if self.num_node_id > 0:
+                # If there are more node ids than the message size, continue on to the next message
+                # Otherwise, locate the index of the data size
+
+                if self.num_node_id >= MESSAGE_SIZE:
+                    self.num_node_id = self.num_node_id - MESSAGE_SIZE
+                else:
+                    bytes_remaining = MESSAGE_SIZE - self.num_node_id
+                    if bytes_remaining == 1:
+                        self.num_data_bytes = message_bytearray[self.num_node_id] << 8
+                        self.incomplete_data_bytes = True
+                    elif bytes_remaining >= 2:
+                        self.num_data_bytes = message_bytearray[self.num_node_id] << 8 | message_bytearray[self.num_node_id + 1]
+                        self.incomplete_data_bytes = False
+
+                        if bytes_remaining > 2:
+                            # Determine the index of the first data byte (If this is 8, then it is
+                            # in the next message)
+                            first_data_index = self.num_node_id
+
+                            # If there are data bytes in this message, subtract that number and
+                            # continue
+                            if first_data_index < MESSAGE_SIZE:
+                                self.num_data_bytes = self.num_data_bytes - \
+                                    (MESSAGE_SIZE - (first_data_index))
+                            # Otherwise, we continue to the next message
+                    self.num_node_id = 0
+            else:
+                # There are no more node ids.
+                # If we only got half of the data bytes
+                bytes_remaining = MESSAGE_SIZE
+                if self.incomplete_data_bytes:
+                    self.num_data_bytes = self.num_data_bytes | message_bytearray[0]
+                    bytes_remaining -= 1
+                # If we already have the data bytes, use how many is remaining (otherwise
+                # it was just set)
+                if self.num_data_bytes > bytes_remaining:
+                    self.num_data_bytes = self.num_data_bytes - bytes_remaining
+                else:
+                    # There are no more data bytes left, message is complete!
+                    self.receiving_datagram = False
+            for byte in msg.data:
+                self.datagram_messages.append(byte)
+
+            # Once we're on the last message, call the callback
+            if not self.receiving_datagram:
+                self.callback(self.datagram_messages)
```

Comment:
IMO you should deserialize the bytearray before passing it into the callback so clients don't have to do that - not sure how you'd want to deal with notifying the user of deserializing errors then, maybe have an optional error callback?

Reply:
My newer implementation handles this in a different way now. I'll move the conversation to the other comment, and look for your feedback there.